core/authentication_api.php

```diff
@@ -717,6 +717,20 @@ function auth_does_password_match( $p_user_id, $p_test_password ) {
 	}
 
 	$t_password = user_get_field( $p_user_id, 'password' );
+
+	# Process SAFE_HASH separately from the legacy methods, to safeguard against
+	# timing attacks leveraging password_verify() function.
+	if(    SAFE_HASH == $t_configured_login_method
+		&& password_verify( $p_test_password, $t_password )
+	) {
+		# Update the password hash if it does not match the current algorithm
+		if( password_needs_rehash( $t_password, PASSWORD_DEFAULT ) ) {
+			user_set_password( $p_user_id, $p_test_password, true );
```

Comment:
I'm not recommending a specific approach, you can make such decision.  I just think we should skip the steps that don't apply when we are changing the hash but the password remains the same.